conbench/entities/distribution.py

```diff
@@ -108,109 +115,122 @@ def get_distribution(benchmark_result, limit):
+    # second benchmark result. But that's too rare and small-impact to justify not
+    # caching for performance reasons.
+    cached_baseline_runs = {}
 
     for benchmark_result in benchmark_results:
         benchmark_result.z_score = None
+        if benchmark_result.error:
+            log.debug("benchmark_result has error; setting z_score = None")
+            continue
 
-    first = benchmark_results[0]
-    parent_commit = get_closest_parent(first.run)
-    if not parent_commit:
-        return
-
-    where = [
-        Distribution.commit_id == parent_commit.id,
-        Distribution.hardware_hash == first.run.hardware.hash,
-    ]
-    if len(benchmark_results) == 1:
-        where.extend(
-            [
-                Distribution.case_id == first.case_id,
-                Distribution.context_id == first.context_id,
-            ]
-        )
+        if benchmark_result.run_id in cached_baseline_runs:
+            baseline_run = cached_baseline_runs[benchmark_result.run_id]
+        else:
+            baseline_run = benchmark_result.run.get_baseline_run(
+                on_default_branch=True,
+                case_id=benchmark_result.case_id,
+                context_id=benchmark_result.context_id,
+            )
+            cached_baseline_runs[benchmark_result.run_id] = baseline_run
 
-    cols = [
-        Distribution.case_id,
-        Distribution.context_id,
-        Distribution.mean_mean,
-        Distribution.mean_sd,
-    ]
+        if not baseline_run:
+            log.debug("No baseline run; setting benchmark_result.z_score = None")
+            continue
 
-    distributions = Session.query(*cols).filter(*where).all()
-    lookup = {f"{d.case_id}-{d.context_id}": d for d in distributions}
+        commit = baseline_run.commit_id
+        hardware = baseline_run.hardware.hash
 
-    for benchmark_result in benchmark_results:
-        if benchmark_result.error:
-            continue
+        if (commit, hardware) in cached_distributions:
+            distributions = cached_distributions[(commit, hardware)]
+        else:
+            distributions = (
+                Session.query(Distribution)
+                .filter_by(commit_id=commit, hardware_hash=hardware)
+                .all()
+            )
+            cached_distributions[(commit, hardware)] = distributions
+
+        # based on the unique index, this list should either have length 0 or 1
+        matching_distributions = [
+            distribution
+            for distribution in distributions
+            if distribution.case_id == benchmark_result.case_id
+            and distribution.context_id == benchmark_result.context_id
+        ]
+
+        if (
+            benchmark_result.mean is not None
+            and matching_distributions
+            and matching_distributions[0].mean_mean is not None
+            and matching_distributions[0].mean_sd  # is positive
```

Comment:
This is a statement about the sd, right? And not a test of positivity?

```suggestion
            and matching_distributions[0].mean_sd  # is always positive
```

Reply:
What I meant was we need to make sure both that it's not None and not 0, and I was using python truthiness to make both of those checks at once. Since you asked the question, it's probably too magic :) I can change it to be more explicit